internal/contentmanager/source/dynamicsource.go

```diff
@@ -0,0 +1,195 @@
+}
+
+func (dis *dynamicInformerSource) Start(ctx context.Context, q workqueue.TypedRateLimitingInterface[reconcile.Request]) error {
+	// Close the startedChan to signal that this
+	// source has been started. Subsequent calls
+	// to Start will attempt to close a closed channel
+	// and panic.
+	close(dis.startedChan)
+
+	dis.informerCtx, dis.informerCancel = context.WithCancel(ctx)
+	gInf := dis.cfg.DynamicInformerFactory.ForResource(dis.cfg.GVR)
+	eventHandler := source.NewEventHandler(dis.informerCtx, q, dis.cfg.Handler, dis.cfg.Predicates)
+
+	// If we encounter an error during the watch we will:
+	// - Capture the error
+	// - cancel the informer
+	// requeuing of the ClusterExtension should happen by the
+	// WaitForSync function returning an error
+	// Only if we have successfully synced in the past should we
+	// requeue the ClusterExtension
+	sharedIndexInf := gInf.Informer()
+	err := sharedIndexInf.SetWatchErrorHandler(func(r *cgocache.Reflector, err error) {
+		dis.errOnce.Do(func() {
+			dis.err = err
+			close(dis.erroredChan)
+		})
+
+		if dis.hasSynced() {
+			// We won't be able to update the ClusterExtension status
+			// conditions so instead force a requeue if we
+			// have previously synced and then errored
+			q.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: dis.cfg.Owner.GetName(),
+				},
+			})
+			dis.cfg.OnPostSyncError(dis.informerCtx)
+		}
+		dis.informerCancel()
+		cgocache.DefaultWatchErrorHandler(r, err)
+	})
+	if err != nil {
+		return fmt.Errorf("setting WatchErrorHandler: %w", err)
+	}
+
+	_, err = sharedIndexInf.AddEventHandler(eventHandler.HandlerFuncs())
+	if err != nil {
+		return fmt.Errorf("adding event handler: %w", err)
+	}
+
+	go sharedIndexInf.Run(dis.informerCtx.Done())
+
+	go func() {
+		syncOnce := sync.OnceFunc(func() {
+			dis.syncedChan <- struct{}{}
+			close(dis.syncedChan)
+		})
+
+		_ = wait.PollUntilContextCancel(dis.informerCtx, time.Second, true, func(ctx context.Context) (bool, error) {
+			if sharedIndexInf.HasSynced() {
+				syncOnce()
+				return true, nil
+			}
+			return false, nil
+		})
+	}()
+
+	return nil
+}
+
+func (dis *dynamicInformerSource) WaitForSync(ctx context.Context) error {
+	if !dis.hasStarted() {
+		return fmt.Errorf("not yet started")
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-dis.erroredChan:
+		return dis.err
+	case <-dis.informerCtx.Done():
+		return dis.informerCtx.Err()
+	case <-dis.syncedChan:
+		return nil
+	}
+}
+
+func (dis *dynamicInformerSource) Close() error {
+	if !dis.hasStarted() {
+		return errors.New("source has not yet started")
+	}
+	dis.informerCancel()
+	return nil
+}
+
+func (dis *dynamicInformerSource) hasSynced() bool {
+	select {
+	case <-dis.syncedChan:
+		return true
+	default:
+		return false
+	}
+}
+
+func (dis *dynamicInformerSource) hasStarted() bool {
+	select {
+	case <-dis.startedChan:
+		return true
+	default:
+		return false
+	}
+}
```

Comment:
Nit: looks like an opportunity for a little reuse (and a slight improvement on checking if the channel is closed (vs having received a value)

```suggestion
func isChannelClosed(ch <-chan struct{}) bool {
	select {
	case _, isOpen <-ch:
	        // we could maybe panic here if we get `isOpen=true`, because that would be an expected receive of a value
		return !isOpen
	default:
		return false
	}
}

func (dis *dynamicInformerSource) hasSynced() bool {
	return isChannelClosed(dis.syncedChan)
}

func (dis *dynamicInformerSource) hasStarted() bool {
	return isChannelClosed(dis.startedChan)
}
```

Reply:
While I think having it as a single function _could_ be nice I'm not sure it buys us all that much. Having distinct functions for `hasSynced()` and `hasStarted()` allows us to modify the behavior as we see fit and gives us flexibility to update the synced vs started logic as needed. 

Going to leave as is for now.